### PR TITLE
Make FSI run on coreclr again

### PR DESCRIPTION
**Pull request by KevinRansom**
FSI was broken on coreclr because:

1. Incorrect usage of LoaderContext.
2. FSI does resolution magic using MSBUILD on Windows Desktop.  When initializing FSI code on coreclr it tried to compute a highest framework version.  But didn't really know how to do it.

This PR: uses LoaderContext correctly.  And works around the initialization issue.

TODO:
A.  Figure out new resolver mechanism that eliminates desktop/MSBUILD style resolver magic.
B.  Eliminate current requirement for command line options.

### CODE REVIEW

File: src/fsharp/CompileOps.fs [hunk 1]
```diff
@@ -1625,10 +1625,6 @@ let DefaultBasicReferencesForOutOfProjectSources =
       // Note: this is not a partiuclarly good technique as it relying on the environment the compiler is executing in
       // to determine the default references. However, System.Core will only fail to load on machines with only .NET 2.0,
       // in which case the compiler will also be running as a .NET 2.0 process.
-      //
-      // NOTE: it seems this can now be removed now that .NET 4.x is minimally assumed when using this toolchain
-      if (try System.Reflection.Assembly.Load(new System.Reflection.AssemblyName("System.Core, Version=3.5.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")) |> ignore; true with _ -> false) then 
-          yield "System.Core"
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: I think you should now _always_  yield System.Core here (remove the condition, not the yield)

File: src/fsharp/CompileOps.fs [hunk 2]
```diff
@@ -2255,7 +2250,7 @@ type TcConfigBuilder =
  #endif
           showTerms     = false 
           writeTermsToFiles = false 
-          
+
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: Remove whitespace changes if possible

File: src/fsharp/CompileOps.fs [hunk 3]
```diff
@@ -2580,7 +2575,11 @@ type TcConfig private (data : TcConfigBuilder,validate:bool) =
             with _ -> 
                 error(Error(FSComp.SR.buildCannotReadAssembly(filename),rangeStartup))
         | _ ->
-            None, MSBuildResolver.HighestInstalledNetFrameworkVersionMajorMinor(), false
+            // TODO:  we have to get msbuild out of this
+            if data.useSimpleResolution then
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: Note Mono uses simple resolution, but does have a correct HighestInstalledNetFrameworkVersionMajorMinor implementation.  So this conditional might need to be if-running-on-coreclr in some way.

File: src/fsharp/CompileOps.fs [hunk 4]
```diff
@@ -4398,10 +4397,10 @@ type TcImports(tcConfigP:TcConfigProvider, initialResolutions:TcAssemblyResoluti
...
                 let foundFile = tcImports.TryResolveAssemblyReference (AssemblyReference (m, speculativeFileName, None), ResolveAssemblyReferenceMode.Speculative)
                 match foundFile with 
-                | OkResult (warns, res) -> 
-                     ReportWarnings warns
-                     tcImports.DoRegisterAndImportReferencedAssemblies(res)
-                     true
+                | OkResult (warns, res) ->
+                    ReportWarnings warns
+                    tcImports.DoRegisterAndImportReferencedAssemblies(res)
+                    true
```
Comment (dsyme) [inline src/fsharp/CompileOps.fs]: More whitespace changes

File: src/fsharp/CompileOps.fs (hunk 3 above)
Comment (KevinRansom) [inline src/fsharp/CompileOps.fs]: I think this whole thing needs to go before too long.  But I will add a #if for now.

File: src/fsharp/CompileOps.fs (hunk 2 above)
Comment (KevinRansom) [inline src/fsharp/CompileOps.fs]: https://github.com/Microsoft/visualfsharp/pull/1407/files?w=1

File: src/utils/reshapedreflection.fs [hunk 5]
```diff
@@ -313,7 +316,13 @@ module internal ReflectionAdapters =
         member this.GetSetMethod() = this.SetMethod
 
 #if FX_RESHAPED_REFLECTION_CORECLR
-    let globalLoadContext = System.Runtime.Loader.AssemblyLoadContext.Default
+    type CustomAssemblyResolver() =
+        inherit AssemblyLoadContext()
+        override this.Load (assemblyName:AssemblyName):Assembly =
+            this.LoadFromAssemblyName(assemblyName)
```
Comment (dsyme) [inline src/utils/reshapedreflection.fs]: Please add a comment on why this overload is needed? Thanks

**Pull request merged**